api/src/main/java/io/grpc/Retainable.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * A <i>Retainable</i> is a resource that can be retained, preventing it being closed too early.
```

Comment:
I'm not wild about how this makes close() a noop and how there is no feedback when the methods are used incorrectly. I think we need an interface that can fail the retain or a way to learn it isn't supported.

I think I was envisioning more a `duplicate()`/`clone()` style of operation that would return a new InputStream. That would keep close() semantics clear. But I'd have to dig deeper to see if that is a pain to implement.

If we want to keep the general approach here, I think we'd want to keep close() doing something even when retained. That could be done by having the semantics be that the inputstream is alive until release()+close(), but in either order. Close() would keep its normal idemponent behavior. But we'd need retain/release to throw instead of being noops if already closed/retained. With this approach it is really weird that you can continue read()ing successfully after close(); seems asking for weird behavior, although I guess this will be used rarely enough to not matter much.

Reply:
Updated to a `duplicate()`/`clone()` style API. See `Detachable`.

In the new API:
- If the Detachable has already been detached, detaching it again would throw.
- Continue reading the detached InputStream behaves as if the InputStream was exhausted; reading the new InputStream behaves as if continuing reading the old InputStream.
- Mark of the original InputStream is discarded after being detached, as the underlying bytes are gone. The new InputStream preserves the mark transferred from the old InputStream.